safetensors/src/tensor.rs

```diff
@@ -388,26 +400,26 @@ impl<'data> SafeTensors<'data> {
     /// The tensor returned is merely a view and the data is not owned by this
     /// structure.
     pub fn tensor(&self, tensor_name: &str) -> Result<TensorView<'data>, SafeTensorError> {
-        if let Some(index) = &self.metadata.index_map.get(tensor_name) {
-            if let Some(info) = &self.metadata.tensors.get(**index) {
-                Ok(TensorView {
-                    dtype: info.dtype,
-                    shape: info.shape.clone(),
-                    data: &self.data[info.data_offsets.0..info.data_offsets.1],
-                })
-            } else {
-                Err(SafeTensorError::TensorNotFound(tensor_name.to_string()))
-            }
-        } else {
-            Err(SafeTensorError::TensorNotFound(tensor_name.to_string()))
-        }
+        let &Some(&index) = &self.metadata.index_map.get(tensor_name) else {
+            return Err(SafeTensorError::TensorNotFound(tensor_name.to_string()));
+        };
+
+        let Some(info) = &self.metadata.tensors.get(index) else {
+            return Err(SafeTensorError::TensorNotFound(tensor_name.to_string()));
+        };
+
+        Ok(TensorView {
+            dtype: info.dtype,
+            shape: info.shape.clone(),
+            data: &self.data[info.data_offsets.0..info.data_offsets.1],
+        })
     }
 
     /// Return the names of the tensors within the SafeTensors.
     /// These are used as keys to access to the actual tensors, that can be
     /// retrieved using the tensor method.
-    pub fn names(&self) -> Vec<&'_ String> {
-        self.metadata.index_map.keys().collect()
+    pub fn names(&self) -> Vec<&'_ str> {
+        self.metadata.index_map.keys().map(String::as_str).collect()
```

Comment:
I'm not sure I see the benefit here, callers can also call it themselves if they really want `str` instead of  `&String` no ?

Reply:
The point here is that `&String` is *strictly* worse than `&str` because:
1. it exposes the implementation detail and forces the implementation to use _exactly_ a `String` (and not, e.g. a `Cow<str>` or an `Arc<str>`), while;
2. not having any benefits to the caller, because `&String` can't do anything more than `&str`. As it's immutable, any functionality that would be useful for `String` manipulation – such as appending characters or increasing the capacity – is inaccessible on the returned reference, so it provides only the functionality of `&str`.
As such, it's preferable to return just an `&str`, because it doesn't take away any functionality from the caller, while allowing the implementation to swap out the backing type if needed.